File: demos/python/error_recovery_and_logging.py
# Copyright 2025 Trossen Robotics
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#    * Redistributions of source code must retain the above copyright
#      notice, this list of conditions and the following disclaimer.
#
#    * Redistributions in binary form must reproduce the above copyright
#      notice, this list of conditions and the following disclaimer in the
#      documentation and/or other materials provided with the distribution.
#
#    * Neither the name of the copyright holder nor the names of its
#      contributors may be used to endorse or promote products derived from
#      this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

# Purpose:
# This script demonstrates how to recover from an error in the driver.

# Hardware setup:
# 1. A WXAI V0 arm with leader end effector and ip at 192.168.1.2

# The script does the following:
# 1. Configures the logging
# 2. Initializes the driver
# 3. Configures the driver
# 4. Sets the arm to position mode
# 5. Moves the arm to the home position
# 6. Triggers an error by setting a joint position to an invalid value
# 7. Recovers from the error
# 8. Moves the arm to the sleep position

from time import sleep

import numpy as np

import trossen_arm

if __name__=='__main__':
    MODEL = trossen_arm.Model.wxai_v0
    SERV_IP = '192.168.1.2'
    USE_LOGURU = False  # Set to False to use Python's built-in logging module

    print("Configuring logging...")
    if USE_LOGURU:
        # Using loguru as an example, but any logging library should work
        # You might need to install loguru in your environment
        from loguru import logger

        # Remove loguru's default stderr sink and add custom sinks
        logger.remove()
        logger.add(
            sink=lambda msg: print(msg, end=""),
            format="[{time:YYYY-MM-DD HH:mm:ss}] [{extra[name]}] [{level}] {message}",
            level="DEBUG",
        )
        logger.add(
            "error_recovery_and_logging.log",
            format="[{time:YYYY-MM-DD HH:mm:ss}] [{extra[name]}] [{level}] {message}",
            level="DEBUG",
        )

        # Map C++ LogLevel to loguru level names
        _LEVEL_MAP = {
            trossen_arm.LogLevel.trace: "TRACE",
            trossen_arm.LogLevel.debug: "DEBUG",
            trossen_arm.LogLevel.info: "INFO",
            trossen_arm.LogLevel.warn: "WARNING",
            trossen_arm.LogLevel.error: "ERROR",
            trossen_arm.LogLevel.critical: "CRITICAL",
        }

        def _loguru_backend(level, name, message):
            loguru_level = _LEVEL_MAP.get(level, "CRITICAL")
            logger.bind(name=name).log(loguru_level, message)

        trossen_arm.TrossenArmDriver.set_logger_backend(_loguru_backend)
    else:
        import logging

        formatter = logging.Formatter(
            '[%(asctime)s] [%(name)s] [%(levelname)s] %(message)s',
            '%Y-%m-%d %H:%M:%S'
        )
        stream_handler = logging.StreamHandler()
        stream_handler.setFormatter(formatter)
        file_handler = logging.FileHandler('error_recovery_and_logging.log')
        file_handler.setFormatter(formatter)
        default_logger = logging.getLogger(trossen_arm.TrossenArmDriver.get_default_logger_name())
        default_logger.setLevel(logging.INFO)
        default_logger.addHandler(stream_handler)
        default_logger.addHandler(file_handler)
        logger = logging.getLogger(
            trossen_arm.TrossenArmDriver.get_logger_name(
                MODEL,
                SERV_IP
            )
        )
        logger.setLevel(logging.INFO)
        logger.addHandler(stream_handler)
        logger.addHandler(file_handler)

    print("Initializing the driver...")
    driver = trossen_arm.TrossenArmDriver()

    print("Configuring the driver...")
    driver.configure(
        MODEL,
        trossen_arm.StandardEndEffector.wxai_v0_leader,
        SERV_IP,
        False
    )

    driver.set_all_modes(trossen_arm.Mode.position)

    sleep_positions = np.array(driver.get_all_positions())
    home_positions = np.zeros(driver.get_num_joints())
    home_positions[1] = np.pi/2
    home_positions[2] = np.pi/2

    try:
        print("Moving the arm to the home position...")
        driver.set_all_positions(home_positions)
        print("Triggering a discontinuity error...")
        # Command a huge step change that the arm cannot physically follow
        # which triggers an error for safety reasons
        home_positions[5] += np.pi
        driver.set_all_positions(home_positions, 0.0)
        sleep(1.0)
        print("Moving the arm to the sleep position...")
        driver.set_all_positions(sleep_positions)
    except Exception as e:
        print("An error occurred: ", e)
        print("Recovering from the error...")

        # Cleanup with reboot_controller = False and configure with clear_error = True
        driver.cleanup()
        driver.configure(
            trossen_arm.Model.wxai_v0,
            trossen_arm.StandardEndEffector.wxai_v0_leader,
            '192.168.1.2',
            True
        )
        # Simply calling driver.clear_error() would accomplish the same thing
        # driver.clear_error()

        driver.set_all_modes(trossen_arm.Mode.position)
        driver.set_all_positions(sleep_positions)

File: demos/cpp/error_recovery_and_logging.cpp
// Copyright 2025 Trossen Robotics
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the copyright holder nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

// Purpose:
// This script demonstrates how to recover from an error in the driver.
//
// Hardware setup:
// 1. A WXAI V0 arm with leader end effector and ip at 192.168.1.2
//
// The script does the following:
// 1. Configures the logging
// 2. Initializes the driver
// 3. Configures the driver
// 4. Sets the arm to position mode
// 5. Moves the arm to the home position
// 6. Triggers an error by setting a joint position to an invalid value
// 7. Recovers from the error
// 8. Moves the arm to the sleep position

#include <cmath>
#include <iostream>
#include <memory>
#include <thread>
#include <vector>

#include <spdlog/sinks/basic_file_sink.h>
#include <spdlog/sinks/stdout_color_sinks.h>
#include <spdlog/spdlog.h>

#include "libtrossen_arm/trossen_arm.hpp"

int main() {
  const trossen_arm::Model MODEL = trossen_arm::Model::wxai_v0;
  const std::string SERV_IP = "192.168.1.2";

  std::cout << "Configuring logging..." << std::endl;

  // Set up spdlog with stdout and file sinks
  auto stdout_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
  auto file_sink = std::make_shared<spdlog::sinks::basic_file_sink_mt>(
    "error_recovery_and_logging.log",
    true
  );
  spdlog::sinks_init_list sinks{stdout_sink, file_sink};

  // Route all library log messages through spdlog using set_logger_backend
  trossen_arm::TrossenArmDriver::set_logger_backend(
    [sinks](
      trossen_arm::LogLevel level,
      const std::string & name,
      const std::string & message)
    {
      auto logger = spdlog::get(name);
      if (!logger) {
        logger = std::make_shared<spdlog::logger>(name, sinks);
        spdlog::register_logger(logger);
      }

      switch (level) {
        case trossen_arm::LogLevel::trace:
          logger->trace(message);
          break;
        case trossen_arm::LogLevel::debug:
          logger->debug(message);
          break;
        case trossen_arm::LogLevel::info:
          logger->info(message);
          break;
        case trossen_arm::LogLevel::warn:
          logger->warn(message);
          break;
        case trossen_arm::LogLevel::error:
          logger->error(message);
          break;
        case trossen_arm::LogLevel::critical:
          logger->critical(message);
          break;
      }
    }
  );

  std::cout << "Initializing the driver..." << std::endl;
  trossen_arm::TrossenArmDriver driver;

  std::cout << "Configuring the driver..." << std::endl;
  driver.configure(
    MODEL,
    trossen_arm::StandardEndEffector::wxai_v0_leader,
    SERV_IP,
    false
  );

  driver.set_all_modes(trossen_arm::Mode::position);

  std::vector<double> sleep_positions = driver.get_all_positions();
  std::vector<double> home_positions(driver.get_num_joints(), 0.0f);
  home_positions[1] = M_PI_2;
  home_positions[2] = M_PI_2;

  try {
    std::cout << "Moving the arm to the home position..." << std::endl;
    driver.set_all_positions(home_positions);

    std::cout << "Triggering a discontinuity error..." << std::endl;
    // Command a huge step change that the arm cannot physically follow
    // which triggers an error for safety reasons
    home_positions[5] += M_PI;
    driver.set_all_positions(home_positions, 0.0f);

    std::this_thread::sleep_for(std::chrono::seconds(1));

    std::cout << "Moving the arm to the sleep position..." << std::endl;
    driver.set_all_positions(sleep_positions);
  } catch (const std::exception &e) {
    std::cout << "An error occurred: " << e.what() << std::endl;
    std::cout << "Recovering from the error..." << std::endl;

    // Cleanup with reboot_controller = false and configure with clear_error = true
    driver.cleanup();
    driver.configure(
      trossen_arm::Model::wxai_v0,
      trossen_arm::StandardEndEffector::wxai_v0_leader,
      "192.168.1.2",
      true
    );
    // Simply calling driver.clear_error() would accomplish the same thing
    // driver.clear_error();

    driver.set_all_modes(trossen_arm::Mode::position);
    driver.set_all_positions(sleep_positions);
  }

  return 0;
}
